software/contrib/arp.py

```diff
@@ -0,0 +1,261 @@
+#!/usr/bin/env python3
+"""Arpeggiator & ascending/descending scale generator
+
+Outputs ascending/descending quantized CV, with additional outputs
+a perfect fifth up/down (following the circle of fifths)
+"""
+
+from europi import *
+from europi_script import EuroPiScript
+
+from experimental.quantizer import CommonScales, SEMITONE_LABELS, VOLTS_PER_OCTAVE, VOLTS_PER_SEMITONE, SEMITONES_PER_OCTAVE
+
+import random
+
+MODE_ASCENDING = 0
+MODE_DESCENDING = 1
+MODE_RANDOM = 2
+
+def shuffle(l):
+    """Shuffle a list randomly
+
+    Replacement for random.shuffle, which doesn't exist in micropython
+
+    @param l  The list to shuffle
+    """
+    for i in range(len(l)):
+        n = random.randint(i, len(l)-1)
+        tmp = l[i]
+        l[i] = l[n]
+        l[n] = tmp
+
+class Arpeggio:
+    def __init__(self, scale, mode):
+        """Create an arpeggio using the notes from the given scale
+
+        @param scale  A quantizer scales whose notes will be used
+        @param mode   One of ascending, descending, or random
+        """
+        self.mode = mode
+        self.change_scale(scale)
+
+    def change_scale(self, scale):
+        """Change the current scale
+
+        @param scale  The new quantizer scale we want to use
+        """
+        self.semitones = [
+            i for i in range(len(scale.notes)) if scale[i]
+        ]
+        if self.mode == MODE_RANDOM:
+            # shuffle so we have a chance to choose the last note first
+            # see @next_node
+            shuffle(self.semitones)
+
+    def next_note(self):
+        """Get the next note that should be played
+
+        @return The semitone as an integer from 0-12
+        """
+        if self.mode == MODE_ASCENDING:
+            semitone = self.semitones.pop(0)
```

Comment:
[discussion] I'm curious why you're modifying the `self.semitones` list instead of using something like a read index pointer to keep track of the current/next note (similar to `self.current_scale_index`)?

Reply:
It seemed easier (i.e. slightly less code to write) to use a the semitones like a circular queue, popping the just-played note off and adding it to the back. With an index you'd need to worry about resetting it to zero, modular math to avoid overflows, etc... They're equivalent in the end, this just seemed easier at the time.